geopandas/tests/test_clip.py

```diff
@@ -0,0 +1,232 @@
+    buffered_locs["geometry"] = buffered_locs.buffer(4)
+    buffered_locs["type"] = "plot"
+    return buffered_locs
+
+
+@pytest.fixture
+def donut_geometry(buffered_locations, single_rectangle_gdf):
+    """ Make a geometry with a hole in the middle (a donut). """
+    donut = gpd.overlay(
+        buffered_locations, single_rectangle_gdf, how="symmetric_difference"
+    )
+    return donut
+
+
+@pytest.fixture
+def two_line_gdf():
+    """ Create Line Objects For Testing """
+    linea = LineString([(1, 1), (2, 2), (3, 2), (5, 3)])
+    lineb = LineString([(3, 4), (5, 7), (12, 2), (10, 5), (9, 7.5)])
+    gdf = gpd.GeoDataFrame([1, 2], geometry=[linea, lineb], crs={"init": "epsg:4326"})
+    return gdf
+
+
+@pytest.fixture
+def multi_poly_gdf(donut_geometry):
+    """ Create a multi-polygon GeoDataFrame. """
+    multi_poly = donut_geometry.unary_union
+    out_df = gpd.GeoDataFrame(
+        geometry=gpd.GeoSeries(multi_poly), crs={"init": "epsg:4326"}
+    )
+    out_df = out_df.rename(columns={0: "geometry"}).set_geometry("geometry")
+    out_df["attr"] = ["pool"]
+    return out_df
+
+
+@pytest.fixture
+def multi_line(two_line_gdf):
+    """ Create a multi-line GeoDataFrame.
+
+    This has one multi line and another regular line.
+    """
+    # Create a single and multi line object
+    multiline_feat = two_line_gdf.unary_union
+    linec = LineString([(2, 1), (3, 1), (4, 1), (5, 2)])
+    out_df = gpd.GeoDataFrame(
+        geometry=gpd.GeoSeries([multiline_feat, linec]), crs={"init": "epsg:4326"}
+    )
+    out_df = out_df.rename(columns={0: "geometry"}).set_geometry("geometry")
+    out_df["attr"] = ["road", "stream"]
+    return out_df
+
+
+@pytest.fixture
+def multi_point(point_gdf):
+    """ Create a multi-point GeoDataFrame. """
+    multi_point = point_gdf.unary_union
+    out_df = gpd.GeoDataFrame(
+        gpd.GeoSeries([multi_point, Point(2, 5), Point(-11, -14), Point(-10, -12)]),
+        crs={"init": "epsg:4326"},
+    )
+    out_df = out_df.rename(columns={0: "geometry"}).set_geometry("geometry")
+    out_df["attr"] = ["tree", "another tree", "shrub", "berries"]
+    return out_df
+
+
+@pytest.fixture
+def mixed_gdf():
+    """ Create a Mixed Polygon and LineString For Testing """
+    line = LineString([(1, 1), (2, 2), (3, 2), (5, 3), (12, 1)])
+    poly = Polygon([(3, 4), (5, 2), (12, 2), (10, 5), (9, 7.5)])
+    gdf = gpd.GeoDataFrame([1, 2], geometry=[line, poly], crs={"init": "epsg:4326"})
+    return gdf
+
+
+def test_not_gdf(single_rectangle_gdf):
+    """Non-GeoDataFrame inputs raise attribute errors."""
+    with pytest.raises(TypeError):
+        cl.clip((2, 3), single_rectangle_gdf)
+    with pytest.raises(TypeError):
+        cl.clip(single_rectangle_gdf, (2, 3))
+
+
+def test_returns_gdf(point_gdf, single_rectangle_gdf):
+    """Test that function returns a GeoDataFrame (or GDF-like) object."""
+    out = cl.clip(point_gdf, single_rectangle_gdf)
+    assert hasattr(out, "geometry")
```

Comment:
```suggestion
    assert isinstance(out, GeoDataFrame)
```

I think we can be explicit here in the test.

Also, if you input a GeoSeries, you get a GeoSeries out? Maybe can add a test for that here as well:

```
out = cl.clip(point_gdf.geometry, single_rectangle_gdf)
assert isinstance(out, GeoSeries)
```

Reply:
@jorisvandenbossche Hey! I updated the existing test to insure that the output is actually a GeoDataFrame. However, the second test you recommend to test that a GeoSeries is returned fails. That's because the clip function in its current state can't handle a GeoSeries. It fails on the last line, with the following error:

```
return concat.sort_values(by="_order").drop(columns="_order")
E       TypeError: sort_values() got an unexpected keyword argument 'by'
```

If this is something we want to change so that it can accept a GeoSeries, let me know!

Reply:
Ah, yes, that's something we would like to fix. It's the patch to ensure the original order that is only working for dataframes, and not for series. 

I can look into making a version that works for series as well

Reply:
Thank you @jorisvandenbossche !!